File: _98_validate_binary_search_tree/main.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return exploreSubtree(root, LONG_MIN, LONG_MAX);
    }

    bool exploreSubtree(TreeNode* root, long minimum, long maximum)
    {
        if (root == NULL)
        {
            return true;
        }

        if (root->val <= minimum || root->val >= maximum)
        {
            return false;
        }

        return exploreSubtree(root->left, minimum, root->val) && exploreSubtree(root->right, root->val, maximum);
    }
};

File: _98_validate_binary_search_tree/main.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.exploreSubtree(root, -math.inf, math.inf)

    def exploreSubtree(self, root: Optional[TreeNode], minimum, maximum) -> bool:
        if not root:
            return True

        if root.val <= minimum or root.val >= maximum:
            return False

        return self.exploreSubtree(root.left, minimum, root.val) and self.exploreSubtree(root.right, root.val, maximum)